pkg/controller/podautoscaler/horizontal.go

```diff
@@ -419,35 +433,67 @@ func (a *HorizontalController) reconcileAutoscaler(hpav1Shared *autoscalingv1.Ho
 			rescaleReason = "All metrics below target"
 		}
 
-		if hpa.Spec.MinReplicas != nil && desiredReplicas < *hpa.Spec.MinReplicas {
-			desiredReplicas = *hpa.Spec.MinReplicas
-		}
+		// Do not upscale too much to prevent incorrect rapid increase of the number of master replicas caused by
+		// bogus CPU usage report from heapster/kubelet (like in issue #32304).
+		scaleUpLimit := calculateScaleUpLimit(currentReplicas)
 
-		//  never scale down to 0, reserved for disabling autoscaling
-		if desiredReplicas == 0 {
+		switch {
+		case desiredReplicas > scaleUpLimit:
+			setCondition(hpa, autoscalingv2.DesiredOutsideRange, v1.ConditionTrue, "ScaleUpLimit", "the desired replica count increasing faster than the maximum scale rate")
```

Comment:
Optional (here and in all cases): consider adding desired replicas values (before and after modification) to the condition message.